### PR TITLE
feat: update `initialize` endpoint and create `assessments/feedback` endpoint in ORA Staff Grader

### PR DESCRIPTION
**Pull request by BryanttV**
For this PR we have the following goals:

1. Create two new endpoints:
    - The `/api/ora_staff_grader/assessments/feedback/from` endpoint to list the assessment received for a specific submission of a learner.
    - The `/api/ora_staff_grader/assessments/feedback/to` endpoint to list the assessment given by a learner based on their submission ID.
2. Update the `/api/ora_staff_grader/initialize` endpoint to add the `fullname` and `email` fields.

> [!NOTE]  
> [This PR](https://github.com/openedx/frontend-app-ora-grading/pull/263) in the ORA Grading MFE will use these endpoints to display a more detailed table to the instructor.

## Dependecies

> [!IMPORTANT]  
> This upgrade will consist of two PRs: one for ORA and the other for edx-platform. This is because the main serializer of the service resides in the platform. To test it, you must fetch and set up the platform version with the serializer modification.
> - **edx-platform PR**: https://github.com/openedx/edx-platform/pull/33632

#### `/api/ora_staff_grader/assessments/feedback/from`

This endpoint fetches assessment details for a given:
- Submission UUID
- ORA Location

#### `/api/ora_staff_grader/assessments/feedback/to`

This endpoint fetches assessment details for a given:
- Submission UUID
- ORA Location

Both endpoints return the scorer information and scoring details to fill the following Staff Grade MFE table:

image

The following is an example of the response from either of the two endpoints:

```json
{
   "assessments": [
      {
         "assessment_id" : "8",
         "scorer_name" : "Nikola Tesla",
         "scorer_username" : "nikola",
         "scorer_email" : "nikola@tesla.com",
         "assesment_date" : "2024-02-15 15:29:55.374959+00:00",
         "assesment_scores" : [
            {
               "criterion_name" : "Ideas",
               "score_earned" : 3,
               "score_type" :" Fair"
            },
            {
               "criterion_name" :" Content",
               "score_earned" : 1,
               "score_type" :" Fair"
            }
         ],
         "problem_step" : "Peer",
         "feedback" : "Good Job"
      }
   ]
}
``` 

#### `/api/ora_staff_grader/initialize`

This is a pre-existing endpoint to which two extra fields are added (`fullname`, `email`) to fill the following table:

image

The following is an example of the response from this endpoint:

```json
...
    "submissions" : {
        "619b8fb2-ebc0-49ed-9b35-4700efb28f66" : {
            "submissionUUID" : "619b8fb2-ebc0-49ed-9b35-4700efb28f66",
            "username" : "admin",
            "email" : "admin@admin.com",
            "fullname" : "Fernando Gonzalez",
            "teamName" : null,
            "dateSubmitted" : "2023-10-13 15:36:13.781179+00:00",
            "dateGraded" : "2023-10-17 20:26:07.146547+00:00",
            "gradedBy" : "admin",
            "gradeStatus" : "graded",
            "lockStatus" : "unlocked",
            "score" : {
                "pointsEarned" : 10,
                "pointsPossible" : 10
            }
        }
    },
``` 

## Developer Checklist

- [ ] Reviewed the [release process](https://github.com/openedx/edx-ora2/blob/master/.github/release_process.md)
- [ ] Translations and JS/SASS compiled
- [ ] Bumped version number in [openassessment/\_\_init\_\_.py](https://github.com/openedx/edx-ora2/blob/master/openassessment/__init__.py#L4) and [package.json](https://github.com/openedx/edx-ora2/blob/master/package.json#L3)

## Testing Instructions

To test this, you can use Postman:

1. Get an auth token from `/oauth2/access_token/` with the following **x-www-form-urlencoded** content type body:

    ```
    client_id: login-service-client-id
    client_secret: <your-client-secret>
    grant_type: password
    username: <your-username>
    password: <your-password>
    ```
    From the response, copy the `access_token`, which must be sent in the headers of the endpoints.

3. Test the endpoints:
- `/api/ora_staff_grader/initialize`
    - **Query Params**: oraLocation
- `/api/ora_staff_grader/assessments/feedback/from`
    - **Query Params**: oraLocation, submissionUUID
- `/api/ora_staff_grader/assessments/feedback/to`
    - **Query Params**: oraLocation, submissionUUID
 
**NOTE**: From the MFE ora-grading you can get the `oraLocation` from here:

image

And the `submissionUUID` from the response of the `initialize` endpoint here:

image

Make sure that the `oraLocation` value is encoded.

image

## Reviewer Checklist

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [ ] I've tested the new functionality

FYI: @openedx/content-aurora

### CODE REVIEW

**Comment by itsjeyd** [comment]
Hey @BryanttV, thank you for this contribution!

I'm lining the changes up for a test run.

Once everything turns green engineering review can start.

**Comment by e0d** [comment]
@BryanttV can you have a look at the test failures?

**Comment by BryanttV** [comment]
Hi @e0d, I've updated the tests, it's ready for review

**Comment by itsjeyd** [comment]
@BryanttV Thanks for the update. Your PR description is a thing of beauty 😄 

@mattcarter This PR is ready for engineering review by Aurora.

**Comment by pomegranited** [comment]
Hi Hey @mattcarter CC @openedx/2u-aurora @itsjeyd 

I'm a core contributor on this repo, so I can review and merge this and the other ORA2 OSPRs, if that helps you and your team? Please let me know.

**Comment by itsjeyd** [comment]
@pomegranited Thanks for offering to help out here! Per [decisions](https://openedx.atlassian.net/wiki/spaces/COMM/pages/4031578215/2024-02-01+Meeting+notes#%E2%A4%B4-Decisions) made at the first meeting of the new maintenance working group, you're good to go ahead with reviewing and merging this PR 🎉

File: openassessment/staffgrader/staff_grader_mixin.py
Comment (pomegranited) [inline openassessment/staffgrader/staff_grader_mixin.py:215]: Hi @BryanttV , thank you for submitting these PRs! I've only started reviewing, and should have a full review ready for you tomorrow.

From the PR description:

> ### Considerations
> In the new /api/ora_staff_grader/assessments/feedback endpoint there is a query parameter that is used to filter the assessment grades based on the type (received or given) for a specific submission.
> 
> For the name of this query parameter we had 2 possible proposals:
> 
>   * AssessmentType: We think this is a clear name, however, from ORA there is already a reference to this name to classify step types such as Self, Staff, Peer, and Training.
>   * AssessmentFilter: There is no reference to this name from ORA, but it may be too generic and does not clearly describe what the function of the query parameter is.
> 
> Does anyone have suggestions for a clearer, more descriptive name for this query parameter?

Ya, that's tricky.. but since your [handler is actually just an "if/else" statement wrapping these two options](https://github.com/openedx/edx-ora2/pull/2101/files#diff-325ab5d2eef948ecf0b39bbcda9b0b2071dd1c38f9a3eafa3d0e178413a52dedR230-R235), would you consider using two separate endpoints, e.g.:

* /api/ora_staff_grader/assessments/feedback/from?submissionUUID=...&oraLocation=...
* /api/ora_staff_grader/assessments/feedback/to?submissionUUID=...&oraLocation=...

Comment (pomegranited) [reply openassessment/staffgrader/staff_grader_mixin.py:215]: > Including scorer information and scoring details to fill the following Staff Grade MFE table:

Is there an MFE somewhere that shows this API in use?

Comment (pomegranited) [reply openassessment/staffgrader/staff_grader_mixin.py:215]: Nudge @BryanttV ^

CC @felipemontoya @mariajgrimaldi

Comment (mariajgrimaldi) [reply openassessment/staffgrader/staff_grader_mixin.py:215]: Oh sorry! I'm just seeing this comment. I like the from/to idea since we're not using complex query parameters. And yes! We're using this PR to feed a table in the ORA MFE: https://github.com/openedx/frontend-app-ora-grading/pull/263

@BryanttV: could you link the ORA MFE PR in the cover letter? Also, what do you think about the suggestion? Does it work alongside the ORA Grading MFE?

@pomegranited: we'll find some time this week to address your comments in detail. Thank you for the review!

Comment (pomegranited) [reply openassessment/staffgrader/staff_grader_mixin.py:215]: @mariajgrimaldi Thank you for your reply! I wanted to make sure this change was still needed before doing any further review, and hadn't heard back from @BryanttV .

Let me know when this is ready for another look?

Comment (BryanttV) [reply openassessment/staffgrader/staff_grader_mixin.py:215]: Hi @pomegranited, Thanks for your comments and suggestions! and sorry for not responding sooner. I'm currently making the necessary changes to make the PRs ready for review.

Comment (pomegranited) [reply openassessment/staffgrader/staff_grader_mixin.py:215]: No worries at all @BryanttV, just drop me a message here when you're ready.

And thank you for the link to that MFE, that will help with testing.

Comment (BryanttV) [reply openassessment/staffgrader/staff_grader_mixin.py]: Hi again @pomegranited, I have already made the refactor to separate the endpoints. I updated the cover letter so you can test the endpoints directly following the instructions.

For the moment, you cannot test using the MFE as it does not yet have the corresponding modifications using the current refactoring. I will be attentive to any questions or comments, thank you very much!

File: openassessment/data.py
Comment (mariajgrimaldi) [inline openassessment/data.py:103]: This is missing a proper description.

File: openassessment/data.py
Comment (mariajgrimaldi) [inline openassessment/data.py:1696]: Should we update the `received` and `given` references?

File: openassessment/data.py
Comment (BryanttV) [reply openassessment/data.py:103]: Done

**Review by pomegranited — CHANGES_REQUESTED** [review]
Hi @BryanttV , this is getting much better, thank you! I tested it with https://github.com/openedx/edx-platform/pull/33632 and it's working great.

This isn't your doing, but I'm a bit dismayed by how huge `openassessment/data.py` has become! But I can see that you're following the conventions laid out there, and refactoring this file needn't be part of this PR. But I think we can reduce some code duplication for what you've added.

To make my suggestions clearer, I pushed them to [this branch](https://github.com/eduNEXT/edx-ora2/compare/bav/staff-grader-table-endpoints...open-craft:edx-ora2:jill/bav/staff-grader-table-endpoints), and have linked to specific commits inline below.

Please feel free to push back on my suggestions, or let me know if you have any questions?

File: openassessment/tests/test_data.py
Comment (mariajgrimaldi) [reply openassessment/tests/test_data.py:1980]: I'm pro mocking methods that interact with other types of systems, which we don't need testing. Eg. databases, APIs. So an in-between would be only mocking the calls that interact with the submissions API, the db like _use_read_replica. But I think there's no need mocking map_anonymized_ids_to_user_data or generate_assessment_data

**Comment by BryanttV** [comment]
Hi, @pomegranited. thank you very much for the suggestions, the code looks much better! Also, I modified the unit tests mocking only the database queries.

The only thing left is to update the translations, which is causing the error in the checks. I'll be working on it.

**Review by pomegranited — APPROVED** [review]
@BryanttV One minor request inline, but I'm happy to merge this once that's addressed. Great work here :)

Could you also push a version bump to `openassessment/__init__.py` and `package.json` so we can do a release after merging this?

:+1:

- [x] I tested this on my devstack using the excellent PR test instructions.
      I also checked that only course staff are allowed to use these new endpoints.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] User-facing strings are extracted for translation

**Review by mariajgrimaldi — APPROVED** [review]
I tested this locally with the latest changes, and it's working as expected. The next step is updating the MFE implementation to match the latest changes. Thank you both for all the hard work!

File: openassessment/data.py
Comment (BryanttV) [reply openassessment/data.py:1645]: I forgot that line, changes made!

**Comment by BryanttV** [comment]
Hi @pomegranited, I made the last changes and bumped the version (The version in the `__init__.py` and `package.json` don't match, both were set to `6.1.0`).

**Comment by pomegranited** [comment]
> both were set to 6.1.0

Perfect, thank you @BryanttV ! Merging this now 😄

**Comment by openedx-webhooks** [comment]
@BryanttV 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-ora2&entry.867055334=https://github.com/openedx/edx-ora2/pull/2101&entry.1484655318=https://github.com/BryanttV&entry.752974735=2023-10-31+20:25&entry.1917517419=2024-02-20+21:08&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
